chronos/apps/http/httptable.nim

```diff
@@ -0,0 +1,182 @@
+#
+#        Chronos HTTP/S case-insensitive non-unique
+#              key-value memory storage
+#             (c) Copyright 2021-Present
+#         Status Research & Development GmbH
+#
+#              Licensed under either of
+#  Apache License, version 2.0, (LICENSE-APACHEv2)
+#              MIT license (LICENSE-MIT)
+import std/[tables, strutils]
+
+{.push raises: [Defect].}
+
+type
+  HttpTable* = object
+    table: Table[string, seq[string]]
+
+  HttpTableRef* = ref HttpTable
+
+  HttpTables* = HttpTable | HttpTableRef
+
+proc `-`(x: uint32): uint32 {.inline.} =
+  (0xFFFF_FFFF'u32 - x) + 1'u32
+
+proc LT(x, y: uint32): uint32 {.inline.} =
+  let z = x - y
+  (z xor ((y xor x) and (y xor z))) shr 31
+
+proc decValue(c: byte): int =
+  let x = uint32(c) - 0x30'u32
+  let r = ((x + 1'u32) and -LT(x, 10))
```

Comment:
The procedure replaces table and procedure which performs lookup for decimal characters, it returns values 0-9 for characters `0`..`9` and returns `-1` for everything else.

Reply:
It looks like the algorithm is expressed in more complicated way to get some property such as constant-time execution, but why is this beneficial here? Are there downsides of having more obvious code?

If we stick with this code, it would be good to have some tests that prove that all `char` values are handled correctly.

Reply:
Such test is present
https://github.com/status-im/nim-chronos/pull/153/files#diff-4d184c6e113bb1e348a3e9d271835351b824a481171afe38d01a856a1bf21cccR765-R769